File: mgindb/handler.py
import json
import asyncio
import websockets
import nest_asyncio
nest_asyncio.apply()

class MginDBClient:
    def __init__(self, host='127.0.0.1', port=6446, username='', password=''):
        self.uri = f"ws://{host}:{port}"
        self.username = username
        self.password = password
        self.websocket = None

    async def connect(self):
        self.websocket = await websockets.connect(self.uri)
        auth_data = {'username': self.username, 'password': self.password}
        await self.websocket.send(json.dumps(auth_data))
        response = await self.websocket.recv()
        if response != "MginDB server connected... Welcome!":
            raise Exception("Failed to authenticate: " + response)

    async def send_command(self, command):
        if not self.websocket or self.websocket.closed:
            await self.connect()
        await self.websocket.send(command)
        return await self.websocket.recv()

    def _send_sync_command(self, *args):
        loop = asyncio.get_event_loop()
        command_str = ' '.join(str(arg) for arg in args if arg is not None)

        # Special handling for JSON data in SET command
        if 'SET' in command_str and '{' in command_str:
            key, json_data = command_str.split(' ', 2)[1:3]
            try:
                # Validate JSON before sending
                parsed_data = json.loads(json_data)
                json_data = json.dumps(parsed_data)
                command_str = f'SET {key} {json_data}'
            except json.JSONDecodeError:
                print("Invalid JSON data provided.")

        if loop.is_running():
            return asyncio.ensure_future(self.send_command(command_str))
        return loop.run_until_complete(self.send_command(command_str))

    async def close(self):
        if self.websocket:
            await self.websocket.close()

    def set(self, key, value):
        return self._send_sync_command('SET', key, value)

    def indices(self, action, key=None, value=None):
        return self._send_sync_command('INDICES', action, key, value)

    def incr(self, key, value):
        return self._send_sync_command('INCR', key, value)

    def decr(self, key, value):
        return self._send_sync_command('DECR', key, value)

    def delete(self, key):
        return self._send_sync_command('DEL', key)

    def query(self, key, query_string=None, options=None):
        return self._send_sync_command('QUERY', key, query_string, options)

    def count(self, key):
        return self._send_sync_command('COUNT', key)

    def schedule(self, action, cron_or_key=None, command=None):
        return self._send_sync_command('SCHEDULE', action, cron_or_key, command)

    def sub(self, key):
        return self._send_sync_command('SUB', key)

    def unsub(self, key):
        return self._send_sync_command('UNSUB', key)

File: mgindb/handler.php
<?php
require_once 'vendor/autoload.php';
use TextWebSocketClient;

class MginDBClient {
    private $uri;
    private $username;
    private $password;
    private $websocket;

    public function __construct($host = '127.0.0.1', $port = 6446, $username = '', $password = '') {
        $this->uri = "ws://$host:$port";
        $this->username = $username;
        $this->password = $password;
        $this->websocket = null;
    }

    public function connect() {
        $this->websocket = new TextWebSocketClient($this->uri);
        $this->websocket->connect();
        $authData = json_encode(['username' => $this->username, 'password' => $this->password]);
        $this->websocket->send($authData);
        $response = $this->websocket->receive();
        if ($response !== "MginDB server connected... Welcome!") {
            throw new Exception("Failed to authenticate: " . $response);
        }
    }

    public function sendCommand($command) {
        if ($this->websocket === null || !$this->websocket->isConnected()) {
            $this->connect();
        }
        $this->websocket->send($command);
        return $this->websocket->receive();
    }

    public function set($key, $value) {
        return $this->sendCommand("SET $key $value");
    }

    public function indices($action, $key = null, $value = null) {
        $command = "INDICES $action" . ($key ? " $key" : '') . ($value ? " $value" : '');
        return $this->sendCommand($command);
    }

    public function incr($key, $value) {
        return $this->sendCommand("INCR $key $value");
    }

    public function decr($key, $value) {
        return $this->sendCommand("DECR $key $value");
    }

    public function delete($key) {
        return $this->sendCommand("DEL $key");
    }

    public function query($key, $queryString = null, $options = null) {
        $command = "QUERY $key" . ($queryString ? " $queryString" : '') . ($options ? " $options" : '');
        return $this->sendCommand($command);
    }

    public function count($key) {
        return $this->sendCommand("COUNT $key");
    }

    public function schedule($action, $cronOrKey = null, $command = null) {
        $cmd = "SCHEDULE $action" . ($cronOrKey ? " $cronOrKey" : '') . ($command ? " $command" : '');
        return $this->sendCommand($cmd);
    }

    public function sub($key) {
        return $this->sendCommand("SUB $key");
    }

    public function unsub($key) {
        return $this->sendCommand("UNSUB $key");
    }

    public function close() {
        if ($this->websocket !== null && $this->websocket->isConnected()) {
            $this->websocket->close();
        }
    }
}
?>